[DRAFT] Increase `MAX_EFFECTIVE_BALANCE` minimal spec change

michaelneuder opened the pull request:
Spec changes associated with https://notes.ethereum.org/@mikeneuder/increase-maxeb

For analysis on the changes see https://notes.ethereum.org/@fradamt/meb-increase-security

Changes:
1. exit queue rate limiting by stake weight 
2. activation queue rate limiting by stake weight
3. is aggregator changes
4. limit top ups to 32 eth to avoid by passing activation queue
5. partial withdrawals depending on 0x02 credential

michaelneuder committed:
mvp

michaelneuder committed:
partial withdrawals

michaelneuder committed:
top ups

michaelneuder committed:
top ups

michaelneuder committed:
top ups

michaelneuder committed:
all top ups capped at 32 ETH

fradamt committed:
simplify top-up

michaelneuder committed:
update get_expected_withdrawals

michaelneuder committed:
Merge branch 'maxeb-pyspec-min-viable' of https://github.com/michaelneuder/consensus-specs into maxeb-pyspec-min-viable

michaelneuder committed:
ceiling function

michaelneuder committed:
remove proposer slashing modification –– permature optimizaiton IMO

michaelneuder committed:
rm MIN_PROPOSER_SLASHING

michaelneuder committed:
balance - ceiling

JustinDrake commented on specs/_features/maxeb_increase/capella.py:
```diff
@@ -3589,11 +3633,12 @@ def get_expected_withdrawals(state: BeaconState) -> Sequence[Withdrawal]:
                 validator_index=validator_index,
                 address=ExecutionAddress(validator.withdrawal_credentials[12:]),
-                amount=balance - MAX_EFFECTIVE_BALANCE,
+                amount=balance - ceiling,
```
I suggest not introducing "ceiling" as a new concept: we already have "min activation balance", "max effective balance", and "excess balance". Maybe replace `get_validator_ceiling()` with `get_validator_excess_balance`

michaelneuder replied:
done!

JustinDrake commented on specs/_features/maxeb_increase/capella.py:
```diff
@@ -3540,24 +3559,49 @@ def has_eth1_withdrawal_credential(validator: Validator) -> bool:
+    return validator.withdrawal_credentials[:1] == COMPOUNDING_WITHDRAWAL_PREFIX
+
+
+def has_withdrawable_credential(validator: Validator) -> bool:
```
Suggest removing this function. One can simply write `or has_compounding_withdrawal_credential(validator)` where relevant.

michaelneuder replied:
done!

JustinDrake commented on specs/_features/maxeb_increase/capella.py:
```diff
@@ -770,7 +772,6 @@ def integer_squareroot(n: uint64) -> uint64:
         y = (x + n // x) // 2
     return x
 
-
```
Suggest keeping those new lines to minimise the diff.

michaelneuder replied:
removed :-)

JustinDrake commented on specs/_features/maxeb_increase/capella.py:
```diff
@@ -2334,9 +2348,14 @@ def get_slot_signature(state: BeaconState, slot: Slot, privkey: int) -> BLSSigna
+    validator = state.validators[validator_index]
     committee = get_beacon_committee(state, slot, index)
-    modulo = max(1, len(committee) // TARGET_AGGREGATORS_PER_COMMITTEE)
+    number_virtual_validators = validator.effective_balance // MIN_ACTIVATION_BALANCE
```
Suggest trying to not introduce a new concept of "virtual validator".

michaelneuder replied:
renamed to `min_balance_increments`!

JustinDrake commented on specs/_features/maxeb_increase/capella.py:
```diff
@@ -1794,10 +1808,10 @@ def apply_deposit(state: BeaconState,
-        increase_balance(state, index, amount)
-
+        top_up = min(amount, MIN_ACTIVATION_BALANCE - state.balances[index])
+        increase_balance(state, index, top_up)
```
consider stuffing on one line :)

michaelneuder replied:
done!

JustinDrake commented on specs/_features/maxeb_increase/capella.py:
```diff
@@ -190,7 +191,8 @@ def floorlog2(x: int) -> uint64:
 MIN_DEPOSIT_AMOUNT = Gwei(1000000000)
-MAX_EFFECTIVE_BALANCE = Gwei(32000000000)
+MAX_EFFECTIVE_BALANCE = Gwei(2048000000000)
+MIN_ACTIVATION_BALANCE = Gwei(32000000000)
```
consider putting `MIN_ACTIVATION_BALANCE` first (because it's the same 32 ETH constant)

michaelneuder replied:
done!

JustinDrake commented on specs/_features/maxeb_increase/capella.py:
```diff
@@ -1021,12 +1022,12 @@ def get_active_validator_indices(state: BeaconState, epoch: Epoch) -> Sequence[V
-    active_validator_indices = get_active_validator_indices(state, get_current_epoch(state))
-    return max(config.MIN_PER_EPOCH_CHURN_LIMIT, uint64(len(active_validator_indices)) // config.CHURN_LIMIT_QUOTIENT)
+    total_balance = get_total_active_balance(state)
+    return max(config.MIN_PER_EPOCH_CHURN_LIMIT, total_balance // config.CHURN_LIMIT_QUOTIENT)
```
can be one line

michaelneuder replied:
done!

JustinDrake commented on specs/_features/maxeb_increase/capella.py:
```diff
@@ -1146,10 +1147,19 @@ def initiate_validator_exit(state: BeaconState, index: ValidatorIndex) -> None:
+    if exit_queue_epoch > max(exit_epochs): 
+        state.exit_queue_churn = Gwei(0)
+    churn_limit = get_validator_churn_limit(state)
+    if state.exit_queue_churn + validator.effective_balance <= churn_limit: # the validator fits within the churn of the current exit_queue_epoch
```
inline comments in Python have two spaces before the `#` (e.g. `churn_limit:  # the validator`)

michaelneuder replied:
done!

JustinDrake commented on specs/_features/maxeb_increase/capella.py:
```diff
@@ -1146,10 +1147,19 @@ def initiate_validator_exit(state: BeaconState, index: ValidatorIndex) -> None:
+        state.exit_queue_churn += validator.effective_balance # the full effective balance of the validator contributes to the churn in the exit queue epoch 
+    else: # the validator does not fit within the churn of the current exit_queue_epoch
+        future_epochs_churn_contribution = validator.effective_balance - (churn_limit - state.exit_queue_churn)
+        exit_queue_epoch += Epoch((future_epochs_churn_contribution + churn_limit - 1) // churn_limit) # (numerator + denominator - 1) // denominator rounds up. 
```
pyspec comments don't end with a full stop

michaelneuder replied:
done!

JustinDrake commented on specs/_features/maxeb_increase/capella.py:
```diff
@@ -1146,10 +1147,19 @@ def initiate_validator_exit(state: BeaconState, index: ValidatorIndex) -> None:
+    else: # the validator does not fit within the churn of the current exit_queue_epoch
+        future_epochs_churn_contribution = validator.effective_balance - (churn_limit - state.exit_queue_churn)
+        exit_queue_epoch += Epoch((future_epochs_churn_contribution + churn_limit - 1) // churn_limit) # (numerator + denominator - 1) // denominator rounds up. 
+        # the validator contributes to the churn in the exit queue epoch, based on how much balance is left over at that point 
```
pyspec comments start with a capital letter

michaelneuder replied:
done!

JustinDrake reviewed (commented):
first pass (super minor comments)

JustinDrake commented on specs/_features/maxeb_increase/capella.py:
```diff
@@ -273,7 +275,7 @@ class Configuration(NamedTuple):
     ETH1_FOLLOW_DISTANCE=uint64(2048),
     EJECTION_BALANCE=Gwei(16000000000),
-    MIN_PER_EPOCH_CHURN_LIMIT=uint64(4),
+    MIN_PER_EPOCH_CHURN_LIMIT=Gwei(1280000000000),
```
Consider not changing this constant, instead just multiplying it by `MIN_ACTIVATION_BALANCE` whereever it is used.

michaelneuder replied:
done!

michaelneuder committed:
justin round 1

michaelneuder committed:
remove virtual

michaelneuder committed:
remove whitespace

michaelneuder committed:
min_balance_incrememts

michaelneuder committed:
min_balance_increments

michaelneuder committed:
compute_weak_subjectivity_period

JustinDrake commented on specs/_features/maxeb_increase/capella.py:
```diff
@@ -3589,11 +3623,12 @@ def get_expected_withdrawals(state: BeaconState) -> Sequence[Withdrawal]:
                 validator_index=validator_index,
                 address=ExecutionAddress(validator.withdrawal_credentials[12:]),
-                amount=balance - MAX_EFFECTIVE_BALANCE,
+                amount=balance - excess,
```
should this be `excess` (instead of `balance - excess`)?

JustinDrake replied:
consider just doing `amount=get_validator_excess_balance(validator, balance)`

michaelneuder replied:
nice catch! done :-)

JustinDrake commented on specs/_features/maxeb_increase/capella.py:
```diff
@@ -598,7 +600,6 @@ class SignedBLSToExecutionChange(Container):
     message: BLSToExecutionChange
     signature: BLSSignature
 
-
```
No need to remove this line :)

michaelneuder replied:
thanks :-) not sure how these keep getting in there.

JustinDrake commented on specs/_features/maxeb_increase/capella.py:
```diff
@@ -3589,11 +3623,12 @@ def get_expected_withdrawals(state: BeaconState) -> Sequence[Withdrawal]:
             ))
             withdrawal_index += WithdrawalIndex(1)
         elif is_partially_withdrawable_validator(validator, balance):
+            excess = get_validator_excess_balance(validator, balance)
```
consider renaming to `excess_balance` (as done in `is_partially_withdrawable_validator`)

michaelneuder replied:
got rid of the temporary variable!

JustinDrake commented on specs/_features/maxeb_increase/capella.py:
```diff
@@ -3540,24 +3558,40 @@ def has_eth1_withdrawal_credential(validator: Validator) -> bool:
-    return has_eth1_withdrawal_credential(validator) and has_max_effective_balance and has_excess_balance
+    excess_balance = get_validator_excess_balance(validator, balance)
+    has_excess_balance = excess_balance > 0
+    return has_excess_balance
```
consider collapsing into just one line `return get_validator_excess_balance(validator, balance) > 0`

michaelneuder replied:
done!

JustinDrake replied:
Not seeing it on my end :) Did you push?

<img width="940" alt="Screenshot 2023-05-29 at 13 32 45" src="https://github.com/michaelneuder/consensus-specs/assets/731743/5259a0f0-dc0a-4af0-a89b-543433de5784">

JustinDrake commented on specs/_features/maxeb_increase/capella.py:
```diff
@@ -2334,9 +2347,14 @@ def get_slot_signature(state: BeaconState, slot: Slot, privkey: int) -> BLSSigna
+    min_balance_increments = validator.effective_balance // MIN_ACTIVATION_BALANCE
+    committee_balance = get_total_balance(state, set(committee))
+    denominator = committee_balance ** min_balance_increments
+    numerator = denominator - (committee_balance -  TARGET_AGGREGATORS_PER_COMMITTEE * MIN_ACTIVATION_BALANCE) ** min_balance_increments
```
closing bracket seems to be missing

michaelneuder replied:
not sure which you are referring to. i think all are accounted for (and my linter does too)

JustinDrake replied:
Oh, my bad! I did notice two spaces here: `  TARGET_AGGREGATORS_PER_COMMITTEE`

michaelneuder replied:
no problem :-)

JustinDrake commented on specs/_features/maxeb_increase/capella.py:
```diff
@@ -1794,10 +1808,9 @@ def apply_deposit(state: BeaconState,
             state.inactivity_scores.append(uint64(0))
     else:
-        # Increase balance by deposit amount
+        # Increase balance by deposit amount, up to MIN_ACTIVATION_BALANCE.
```
no final `.`

michaelneuder replied:
thanks!

JustinDrake commented on specs/_features/maxeb_increase/capella.py:
```diff
@@ -1547,9 +1557,13 @@ def process_registry_updates(state: BeaconState) -> None:
     ], key=lambda index: (state.validators[index].activation_eligibility_epoch, index))
-    # Dequeued validators for activation up to churn limit
-    for index in activation_queue[:get_validator_churn_limit(state)]:
+    # Dequeue validators for activation up to churn limit [MODIFIED TO BE WEIGHT-SENSITIVE]
```
consider removing ` [MODIFIED TO BE WEIGHT-SENSITIVE]`

JustinDrake replied:
consider also keeping the typo `Dequeued` (vs Dequeue) to reduce the diff

michaelneuder replied:
done! not sure `# Dequeued validators for activation up to churn limit` (the original) is even wrong tbh. will leave as is for now.

JustinDrake commented on specs/_features/maxeb_increase/capella.py:
```diff
@@ -1547,9 +1557,13 @@ def process_registry_updates(state: BeaconState) -> None:
-    # Dequeued validators for activation up to churn limit
-    for index in activation_queue[:get_validator_churn_limit(state)]:
+    # Dequeue validators for activation up to churn limit [MODIFIED TO BE WEIGHT-SENSITIVE]
+    max_churn_left = get_validator_churn_limit(state) # This is now a Gwei amount
```
Consider removing `# This is now a Gwei amount`

michaelneuder replied:
done!

JustinDrake commented on specs/_features/maxeb_increase/capella.py:
```diff
@@ -1146,10 +1147,19 @@ def initiate_validator_exit(state: BeaconState, index: ValidatorIndex) -> None:
+        state.exit_queue_churn = Gwei(0)
+    churn_limit = get_validator_churn_limit(state)
+    if state.exit_queue_churn + validator.effective_balance <= churn_limit:  # The validator fits within the churn of the current exit_queue_epoch
+        state.exit_queue_churn += validator.effective_balance # The full effective balance of the validator contributes to the churn in the exit queue epoch 
```
inline comments need two spaces before `#`

michaelneuder replied:
thanks!

JustinDrake commented on specs/_features/maxeb_increase/capella.py:
```diff
@@ -1547,9 +1557,13 @@ def process_registry_updates(state: BeaconState) -> None:
+    for index in activation_queue:
         validator = state.validators[index]
+        max_churn_left -= validator.effective_balance
+        if max_churn_left < 0:
```
avoid using negative numbers—rewrite as `if max_churn_left < validator.effective_balance`

michaelneuder replied:
reordered to 

```python
for index in activation_queue:
        validator = state.validators[index]
        if max_churn_left < validator.effective_balance:
            break
        max_churn_left -= validator.effective_balance
        validator.activation_epoch = compute_activation_exit_epoch(get_current_epoch(state))
```

JustinDrake reviewed (commented):
minor comments round 2 :)

michaelneuder committed:
justin rd 2 comments

JustinDrake commented on specs/_features/maxeb_increase/capella.py:
```diff
@@ -656,6 +658,7 @@ class BeaconState(Container):
     # Registry
     validators: List[Validator, VALIDATOR_REGISTRY_LIMIT]
     balances: List[Gwei, VALIDATOR_REGISTRY_LIMIT]
+    exit_queue_churn: Gwei
```
I'm expecting exits and activations to be symmetric. If we need `exit_queue_churn` don't we also need `activation_queue_churn`? :)

michaelneuder replied:
@fradamt wrote this part, but IIUC the reason we need this is because we need to preserve the state of the amount of Gwei in the exit_queue epoch has been consumed. for the activation queue, we can just use a constant and fill up as many validators as fit, without worrying about a validator being split over multiple epochs.

dapplion replied:
This variable represent the excess churn of the current highest exit_epoch in the validators records. I think the current name is misleading since exit_queue_churn exist semantically in the codebase and refers to a different thing. I would add `excess` in the name to clarify that difference
```suggestion
    exit_queue_excess_churn: Gwei
```

michaelneuder committed:
justin rd 2 comments

JustinDrake commented on specs/_features/maxeb_increase/capella.py:
```diff
@@ -1794,10 +1809,9 @@ def apply_deposit(state: BeaconState,
+        # Increase balance by deposit amount, up to MIN_ACTIVATION_BALANCE
         index = ValidatorIndex(validator_pubkeys.index(pubkey))
-        increase_balance(state, index, amount)
-
```
keep this new line :)

michaelneuder replied:
im pretty sure my editor is somehow auto-spacing some of these lol. sorry about that.

JustinDrake commented on specs/_features/maxeb_increase/capella.py:
```diff
@@ -3540,24 +3559,38 @@ def has_eth1_withdrawal_credential(validator: Validator) -> bool:
 
+def get_validator_excess_balance(validator: Validator, balance: Gwei) -> Gwei:
+    """
+    Get excess balance for paritial withdrawals for ``validator``.
```
paritial -> partial

michaelneuder replied:
thanks!

JustinDrake commented on specs/_features/maxeb_increase/capella.py:
```diff
@@ -1146,10 +1148,19 @@ def initiate_validator_exit(state: BeaconState, index: ValidatorIndex) -> None:
+    if exit_queue_epoch > max(exit_epochs): 
+        state.exit_queue_churn = Gwei(0)
+    churn_limit = get_validator_churn_limit(state)
+    if state.exit_queue_churn + validator.effective_balance <= churn_limit:  # The validator fits within the churn of the current exit_queue_epoch
+        state.exit_queue_churn += validator.effective_balance  # The full effective balance of the validator contributes to the churn in the exit queue epoch 
+    else:  # The validator does not fit within the churn of the current exit_queue_epoch
+        future_epochs_churn_contribution = validator.effective_balance - (churn_limit - state.exit_queue_churn)
+        exit_queue_epoch += Epoch((future_epochs_churn_contribution + churn_limit - 1) // churn_limit)  # (numerator + denominator - 1) // denominator rounds up
+        # The validator contributes to the churn in the exit queue epoch, based on how much balance is left over at that point 
+        if future_epochs_churn_contribution % churn_limit == 0:
+            state.exit_queue_churn = churn_limit
+        else:
+            state.exit_queue_churn = future_epochs_churn_contribution % churn_limit
```
This whole chunk of code may be significantly more readable with a `while` loop. (Fewer edge cases, potentially no need for fancy rounding trick.)

michaelneuder replied:
i refactored it. basically the idea is now that `get_validator_churn_limit(state)` can never be less than the MAX_EFFECTIVE_BALANCE, we can use the fact that adding one to the exit_queue_epoch guarantees that the validator can exit in a single epoch. this is a lot closer to the logic of before and i think much easier to reason about. the only downside is it requires the churn to be at least MAX_EFFECTIVE_BALANCE, which might not be ok. but at least it makes it easier to present in the current form.

JustinDrake commented on specs/_features/maxeb_increase/capella.py:
```diff
@@ -3540,24 +3559,38 @@ def has_eth1_withdrawal_credential(validator: Validator) -> bool:
+        return balance - MAX_EFFECTIVE_BALANCE
+    elif has_eth1_withdrawal_credential(validator) and balance > MIN_ACTIVATION_BALANCE:
+        return balance - MIN_ACTIVATION_BALANCE
+    return 0
```
Consider `return Gwei(0)` instead of `return 0`. No exactly sure what the rule is. There may be a type linter/checker—may be worth asking Hsiao Wei :)

michaelneuder replied:
done! thanks

JustinDrake commented on specs/_features/maxeb_increase/capella.py:
```diff
@@ -1548,8 +1559,12 @@ def process_registry_updates(state: BeaconState) -> None:
+    max_churn_left = get_validator_churn_limit(state)
+    for index in activation_queue:
         validator = state.validators[index]
+        if max_churn_left < validator.effective_balance:
```
What happens if `validator.effective_balance` is 2048 ETH and `get_validator_churn_limit(state)` is less than 2048 ETH? Does this validator never activate?

michaelneuder replied:
great catch, so the min value of `get_validator_churn_limit` is 128 ETH. let `x` denote the total active balance, then

```
if x / 65536 < 128
```

then the churn limit is 128 ETH meaning no validators will be activated if there is a 2048 validator in the actication queue. it feels like we need get_validator_churn_limit to have a minimum value of MAX_EFFECTIVE_BALANCE, because we always need some way of activating at least 1 validator, unless we want to split large validators over the next few epochs of activation (similar to what we do for withdrawals). I think for now starting with the `MAX_EFFECTIVE_BALANCE` solution is simpler and what ill go with, though this may be part of the design space we iterate on.

michaelneuder replied:
i resolved this by making `config.MIN_PER_EPOCH_CHURN_LIMIT * MIN_ACTIVATION_BALANCE == MAX_EFFECTIVE_BALANCE`, so  changing `MIN_PER_EPOCH_CHURN_LIMIT` 4 to 64 -> a 16x increase. not sure if this is going to fly, but it helps keep the diff reasonable.

michaelneuder committed:
new initiate_validator_exit logic

michaelneuder committed:
new activation and exit logic

michaelneuder commented:
Ok, refactor the activation and exit queue logic based on some discussions with francesco and justin.

### exit queue 

there is a new value in the beacon state called `exit_queue_churn` which keeps track of how much gwei has been accounted for in the current exit epoch. below is the new logic, which ill present in chunks.

```python
    exit_balance_to_consume = validator.effective_balance
    per_epoch_churn_limit = get_validator_churn_limit(state)
    if state.exit_queue_churn + exit_balance_to_consume <= per_epoch_churn_limit:
        state.exit_queue_churn += exit_balance_to_consume
    else:  # Exit balance rolls over to subsequent epoch(s)
        exit_balance_to_consume -= (per_epoch_churn_limit - state.exit_queue_churn)
        exit_queue_epoch += Epoch(1)
        while exit_balance_to_consume >= per_epoch_churn_limit:
            exit_balance_to_consume -= per_epoch_churn_limit
            exit_queue_epoch += Epoch(1)
        state.exit_queue_churn = exit_balance_to_consume
```

first 

```python
    exit_balance_to_consume = validator.effective_balance
    per_epoch_churn_limit = get_validator_churn_limit(state)
```

`exit_balance_to_consume` is the local variable we will use to see how much balance needs to be exited, while `per_epoch_churn_limit` remains the max balance that can exit per epoch.

```python
    if state.exit_queue_churn + exit_balance_to_consume <= per_epoch_churn_limit:
        state.exit_queue_churn += exit_balance_to_consume
```

if sum of `state.exit_queue_churn` (the amount of gwei already exited this epoch) and `exit_balance_to_consume` (the validators effective balance at this point) is less than the churn limit, then we can use `exit_queue_epoch` as the exit epoch for the validator after increasing the `state.exit_queue_churn` by the amount of gwei exiting.

next:
```python
    else:  # Exit balance rolls over to subsequent epoch(s)
        exit_balance_to_consume -= (per_epoch_churn_limit - state.exit_queue_churn)
        exit_queue_epoch += Epoch(1)
```
in this case, we consume the delta between the `per_epoch_churn_limit` and the `state.exit_queue_churn` and increment to the next epoch.

lastly:
```python
    while exit_balance_to_consume >= per_epoch_churn_limit:
        exit_balance_to_consume -= per_epoch_churn_limit
        exit_queue_epoch += Epoch(1)
    state.exit_queue_churn = exit_balance_to_consume
```
we iterate as many epochs as necessary to consume the reset of the balance, setting `state.exit_queue_churn = exit_balance_to_consume` at the end to indicate how much gwei we consumed in the final epoch.

JustinDrake commented on specs/_features/maxeb_increase/capella.py:
```diff
@@ -656,6 +658,8 @@ class BeaconState(Container):
     # Registry
     validators: List[Validator, VALIDATOR_REGISTRY_LIMIT]
     balances: List[Gwei, VALIDATOR_REGISTRY_LIMIT]
+    activation_validator_balance: Gwei
```
Suggest keeping things symmetric, e.g. naming it `activation_queue_churn` to mirror `exit_queue_churn`.

michaelneuder replied:
so the semantics here are slightly different. `exit_queue_churn` tracks the amount of Gwei exited for the current max exit epoch; `activation_validator_balance` tracks the amount of Gwei that the current validator at the front of the activation queue has activated. 

it feels to me like they are quite different values, so naming them to mirror eachother might be less accurate ? wdyt

dapplion replied:
nitpick, but new state variables should be appended at the end to not break proof paths unnecessarily

michaelneuder commented:
### activation queue

the new logic is 

```python
    activation_balance_to_consume = get_validator_churn_limit(state)
    for index in activation_queue:
        validator = state.validators[index]
        # Validator can now be activated
        if state.activation_validator_balance + activation_balance_to_consume >= validator.effective_balance:
            activation_balance_to_consume -= (validator.effective_balance - state.activation_validator_balance)
            state.activation_validator_balance = Gwei(0)
            validator.activation_epoch = compute_activation_exit_epoch(get_current_epoch(state))
        else:  
            state.activation_validator_balance += activation_balance_to_consume
            break
```

we start with a new local variable `activation_balance_to_consume = get_validator_churn_limit(state)` which gets the churn limit for the current epoch. we also introduce a new beacon state variable which keeps track of how much of the current validator balance has been activated so far `state.activation_validator_balance`. 

iterating through each validator we check 

```python
if state.activation_validator_balance + activation_balance_to_consume >= validator.effective_balance:
```

if this is true, the validator "fits" in the current epoch and we can activate them. 

```python
     activation_balance_to_consume -= (validator.effective_balance - state.activation_validator_balance)
     state.activation_validator_balance = Gwei(0)
```
we "consume" the difference between their effective balance and how much has already been activated (could be 0) and then reset `state.activation_validator_balance` to 0 and continue to the next validator.

the else case handles when we cannot fit the validator balance in the current epoch
```python
    else:  
        state.activation_validator_balance += activation_balance_to_consume
        break
```

here we just increment the amount of their balance that has been activated and break out of the loop.

JustinDrake commented on specs/_features/maxeb_increase/capella.py:
```diff
@@ -1548,9 +1558,17 @@ def process_registry_updates(state: BeaconState) -> None:
     # Dequeued validators for activation up to churn limit
-    for index in activation_queue[:get_validator_churn_limit(state)]:
+    activation_balance_to_consume = get_validator_churn_limit(state)
+    for index in activation_queue:
         validator = state.validators[index]
-        validator.activation_epoch = compute_activation_exit_epoch(get_current_epoch(state))
+        # Validator can now be activated
+        if state.activation_validator_balance + activation_balance_to_consume >= validator.effective_balance:
+            activation_balance_to_consume -= (validator.effective_balance - state.activation_validator_balance)
+            state.activation_validator_balance = Gwei(0)
+            validator.activation_epoch = compute_activation_exit_epoch(get_current_epoch(state))
+        else:  
+            state.activation_validator_balance += activation_balance_to_consume
+            break
```
Can this be simplified to the following?

```python
    state.activation_churn_credit += get_validator_churn_limit(state)
    for index in activation_queue:
        validator = state.validators[index]
        if state.activation_churn_credit >= validator.effective_balance:
            state.activation_churn_credit -= validator.effective_balance
            validator.activation_epoch = compute_activation_exit_epoch(get_current_epoch(state))
```

fradamt replied:
There’s a problem with this. Suppose that each epoch has a single 32 ETH validator in the activation queue. Then, state.activation_churn_credit is increased by (churn_limit - 32) per epoch. Later on, state.activation_churn_credit can be so large that a significant fraction of the validator set can exit at once.

michaelneuder replied:
nice! ya this is cleaner. i left the break still to enforce the ordering. working on a set of test cases for these functions now!

JustinDrake replied:
@michaelneuder Francesco is right, the logic is wrong 🙈

michaelneuder replied:
aha, good point! rolled it back to my previous version :-)

michaelneuder committed:
justin rd 3 :-)

michaelneuder committed:
rollback

dapplion commented on specs/_features/maxeb_increase/capella.py:
```diff
@@ -1146,10 +1149,17 @@ def initiate_validator_exit(state: BeaconState, index: ValidatorIndex) -> None:
+        exit_balance_to_consume -= (per_epoch_churn_limit - state.exit_queue_churn)
         exit_queue_epoch += Epoch(1)
-
+        while exit_balance_to_consume >= per_epoch_churn_limit:
+            exit_balance_to_consume -= per_epoch_churn_limit
+            exit_queue_epoch += Epoch(1)
+        state.exit_queue_churn = exit_balance_to_consume
```
The while loop here looks unnecessary, why not do a integer division and modulus operation instead?
```suggestion
        additional_epochs, state.exit_queue_churn = divmod(exit_balance_to_consume - (per_epoch_churn_limit - state.exit_queue_churn), per_epoch_churn_limit)
        exit_queue_epoch += Epoch(additional_epochs + 1)
```

michaelneuder replied:
good call! this is clearer thanks :-)

dapplion commented on specs/_features/maxeb_increase/capella.py:
```diff
@@ -1021,12 +1025,11 @@ def get_active_validator_indices(state: BeaconState, epoch: Epoch) -> Sequence[V
     """
-    active_validator_indices = get_active_validator_indices(state, get_current_epoch(state))
-    return max(config.MIN_PER_EPOCH_CHURN_LIMIT, uint64(len(active_validator_indices)) // config.CHURN_LIMIT_QUOTIENT)
+    return max(config.MIN_PER_EPOCH_CHURN_LIMIT * MIN_ACTIVATION_BALANCE, get_total_active_balance(state) // config.CHURN_LIMIT_QUOTIENT)
```
Looking at the engineering on this on Lodestar, the churn not being align with EFFECTIVE_BALANCE_INCREMENT would break multiple optimizations we have. Due to JS 53 bit representation of stack integers we treat account weights denominated in EFFECTIVE_BALANCE_INCREMENT instead of GWei so the aggregated network balances fit. Aligning the churn to that value should produce a marginal accounting difference in the overall churn so I would really push for this extra line
```suggestion
    churn = max(config.MIN_PER_EPOCH_CHURN_LIMIT * MIN_ACTIVATION_BALANCE, get_total_active_balance(state) // config.CHURN_LIMIT_QUOTIENT)
    return churn - churn % EFFECTIVE_BALANCE_INCREMENT
```

michaelneuder replied:
done!

dapplion commented:
Adding a note for here or the security doc that the weak subjectivity fns need updating https://github.com/michaelneuder/consensus-specs/blob/dev/specs/phase0/weak-subjectivity.md#compute_weak_subjectivity_period

dapplion commented:
To incentivize the network into consolidate the network base reward could be decreased inversely proportional to active index count. Vitalik proposed mechanisms here to cap the set, but placing limits of the _stake_ instead of the _index count_ https://notes.ethereum.org/@vbuterin/single_slot_finality#Economic-capping-of-total-deposits

```python
def get_base_reward_per_increment(state: BeaconState) -> Gwei:
    active_validator_indices_count = len(get_active_validator_indices(state))
    return Gwei(
      (min(active_validator_indices_count, PARTICIPANT_INDEX_COUNT_LIMIT) / active_validator_indices_count)
      * EFFECTIVE_BALANCE_INCREMENT
      * BASE_REWARD_FACTOR // integer_squareroot(get_total_active_balance(state))
    )
```

This way we can push for a healthier state size while letting the market decide the total amount of ETH staked.

At index counts greater than `PARTICIPANT_INDEX_COUNT_LIMIT` it's unreasonable to consider that each key is represented by a different entity, such that consolidation to below that value must be possible. However, there's a common good problem where the entities that choose the consolidate will be penalized via exit and activation queue induced offline periods. Thus the delta achieved by consolidating should compensate for it

michaelneuder committed:
update with pending deposits rate limited by churn

michaelneuder committed:
make pending balance deposit a list

michaelneuder committed:
dapplion response

michaelneuder committed:
small fix

michaelneuder committed:
cache compute_activation_exit_epoch

dapplion commented on specs/_features/maxeb_increase/capella.py:
```diff
@@ -1787,7 +1810,7 @@ def apply_deposit(state: BeaconState,
         if bls.Verify(pubkey, signing_root, signature):
-            state.validators.append(get_validator_from_deposit(pubkey, withdrawal_credentials, amount))
+            state.validators.append(get_validator_from_deposit(pubkey, withdrawal_credentials))
             state.balances.append(amount)
```
Balances are credited immediately, the new deposit and top-up flow is not unified

michaelneuder committed:
custom ceilings

michaelneuder committed:
dapplion response

michaelneuder committed:
remove duplicate is_partially_withdrawable_validator

michaelneuder committed:
adding topups that modify the balance ceiling

michaelneuder committed:
no ceilings

mkalinin commented on specs/_features/maxeb_increase/capella.py:
```diff
@@ -1146,10 +1156,14 @@ def initiate_validator_exit(state: BeaconState, index: ValidatorIndex) -> None:
+        state.exit_queue_churn += exit_balance_to_consume
+    else:  # Exit balance rolls over to subsequent epoch(s)
+        exit_balance_to_consume -= (per_epoch_churn_limit - state.exit_queue_churn)
+        additional_epochs, state.exit_queue_churn = divmod(exit_balance_to_consume - (per_epoch_churn_limit - state.exit_queue_churn), per_epoch_churn_limit)
```
The exit churn leftover has already been subtracted from `exit_balance_to_consume` in the above line, shouldn't this computation be as follows?
```suggestion
        additional_epochs, state.exit_queue_churn = divmod(exit_balance_to_consume, per_epoch_churn_limit)
```

IMHO, fancy Python stuff hinders readability, maybe break this line down into two separate computations?

michaelneuder replied:
removed this in the latest version!

michaelneuder committed:
pending withdrawals

michaelneuder committed:
pull exit queue stuff

michaelneuder committed:
remove old state variable

michaelneuder committed:
adding withdrawal delay

michaelneuder committed:
get_expected_withdrawals

michaelneuder committed:
is_complete

michaelneuder committed:
withdrawable_epoch

michaelneuder committed:
withdrawable_epoch

michaelneuder committed:
FAR_FUTURE_EPOCH

michaelneuder committed:
get_expected_withdrawals

michaelneuder committed:
remove unused local

fradamt committed:
fix typos

fradamt committed:
- Fix typos and small errors
- Make exits withdraw the full balance instead of whater was available when exit is initiated
- Adjust withdrawal amount to be at most balance - 32
- Adjust withdrawable epoch for partial withdrawal when processed after slashing

mkalinin commented on specs/_features/maxeb_increase/capella.py:
```diff
@@ -1575,6 +1582,44 @@ def process_eth1_data_reset(state: BeaconState) -> None:
+    for pending_balance_withdrawal in state.pending_balance_withdrawals:
+        if pending_balance_withdrawal.withdrawable_epoch != FAR_FUTURE_EPOCH:
+            continue
+        if state.withdrawal_validator_balance + withdrawal_balance_to_consume >= pending_balance_withdrawal.amount:
```
If `is_exit` the `amount` should rather be `balances[pending_balance_withdrawal.index]`

mkalinin commented on specs/_features/maxeb_increase/capella.py:
```diff
@@ -1575,6 +1582,44 @@ def process_eth1_data_reset(state: BeaconState) -> None:
+            else:
+                partially_withdrawable_balance = state.balances[validator.index] - MIN_ACTIVATION_BALANCE
+                pending_balance_withdrawal.amount = min(partially_withdrawable_balance, pending_balance_withdrawal.amount)
+            decrease_balance(state, validator.index, pending_balance_withdrawal.amount)
```
The decrease must happen in `process_withdrawals`, not here

michaelneuder committed:
cancel pending withdrawals

dapplion committed:
Add process_execution_layer_withdraw_request

mkalinin committed:
Handle slashings in exit withdrawal queue

mkalinin committed:
Remove unprocessed exit upon slashing

mkalinin committed:
Do not break if withdrawable epoch is not yet come

michaelneuder committed:
Merge pull request #5 from michaelneuder/exit-withdrawal-queue

Updates of pending withdrawals queue processing

michaelneuder committed:
two queues

michaelneuder committed:
two queues

michaelneuder committed:
three queue:

michaelneuder committed:
three queue

mkalinin committed:
Implement exit churn with no physical queue, add partial withdrawals

mkalinin committed:
Fix consumed update

mkalinin committed:
Fix expected withdrawals

mkalinin committed:
Polish names

mkalinin committed:
Skip partial withdrawals if validator exited

mkalinin committed:
Remove unused fields and functions

mkalinin committed:
Cleanup get_expected_withdrawals

mkalinin committed:
Limit partial withdrawals to MAX_WITHDRAWALS_PER_PAYLOAD // 2

fradamt commented on specs/_features/maxeb_increase/capella.py:
```diff
@@ -3576,24 +3720,60 @@ def get_expected_withdrawals(state: BeaconState) -> Sequence[Withdrawal]:
+            break
+
+        validator = state.validators[withdrawal.index]
+
```
I would add `if validator.slashed: continue`, so that withdrawals in `state.eligible_balance_withdrawals` are skipped when there is already a slashed exit to process in `state.eligible_slashing_withdrawals`. This way, it is not necessary to cancel them upon slashing

michaelneuder replied:
i think this is no longer relevant, because we skip and validator with exit_epoch set, which will happen when we initiate validator exit.

fradamt replied:
yeah, it was about the multi-queue version

fradamt commented on specs/_features/maxeb_increase/capella.py:
```diff
@@ -1162,8 +1194,11 @@ def slash_validator(state: BeaconState,
+    validator.exit_epoch = FAR_FUTURE_EPOCH
+    initiate_validator_exit(state, slashed_index)
     validator.slashed = True
     validator.withdrawable_epoch = max(validator.withdrawable_epoch, Epoch(epoch + EPOCHS_PER_SLASHINGS_VECTOR))
```
This assumes that `validator.withdrawable_epoch` is set, which used to happen in `initiate_validator_exit`, but isn't the case here.

michaelneuder replied:
i think this is resolved in the latest ?

michaelneuder committed:
Merge pull request #6 from michaelneuder/queueless-exit-churn

Queueless exit churn

michaelneuder committed:
compute_exit_epoch_and_update_churn tweak

michaelneuder committed:
fix consumed increment indentation level

michaelneuder committed:
add decrease_balance

michaelneuder committed:
rm decrease_balance

mkalinin committed:
Revert changes in slash_validator

fradamt committed:
remove unused field, `withdrawal_balance_to_consume`

dapplion committed:
Constant penalty for proposer equivocations

dapplion committed:
Split slashable status

dapplion committed:
Add is_attester_slashable_validator

mkalinin commented on specs/_features/maxeb_increase/capella.py:
```diff
@@ -2334,9 +2421,14 @@ def get_slot_signature(state: BeaconState, slot: Slot, privkey: int) -> BLSSigna
     committee = get_beacon_committee(state, slot, index)
-    modulo = max(1, len(committee) // TARGET_AGGREGATORS_PER_COMMITTEE)
+    min_balance_increments = validator.effective_balance // MIN_ACTIVATION_BALANCE
+    committee_balance = get_total_balance(state, set(committee))
+    denominator = committee_balance ** min_balance_increments
+    numerator = denominator - (committee_balance -  TARGET_AGGREGATORS_PER_COMMITTEE * MIN_ACTIVATION_BALANCE) ** min_balance_increments
```
```suggestion
    committee_balance_increments = get_total_balance(state, set(committee)) // MIN_ACTIVATION_BALANCE
    denominator = committee_balance_increments ** min_balance_increments
    numerator = denominator - (committee_balance_increments -  TARGET_AGGREGATORS_PER_COMMITTEE) ** min_balance_increments
```

tuanggo replied:
open

dapplion committed:
Don't allow is_slashed_attester to propose blocks

tuanggo reviewed (commented):
ok

michaelneuder committed:
Merge pull request #7 from dapplion/proposer-equivocation-penalty

Constant penalty for proposer equivocations

michaelneuder committed:
Update specs/_features/maxeb_increase/capella.py

Co-authored-by: Mikhail Kalinin <noblesse.knight@gmail.com>

mkalinin committed:
Update EB processing on epoch boundary

mkalinin committed:
Update specs/_features/maxeb_increase/capella.py

Co-authored-by: Lion - dapplion <35266934+dapplion@users.noreply.github.com>

michaelneuder committed:
Merge pull request #11 from michaelneuder/mkalinin-patch-1

Update EB processing on epoch boundary

potuz commented on specs/_features/maxeb_increase/capella.py:
```diff
@@ -680,6 +705,8 @@ class BeaconState(Container):
     next_withdrawal_validator_index: ValidatorIndex  # [New in Capella]
     # Deep history valid from Capella onwards
     historical_summaries: List[HistoricalSummary, HISTORICAL_ROOTS_LIMIT]  # [New in Capella]
+    pending_balance_deposits: List[PendingBalanceDeposit]
+    pending_partial_withdrawals: List[PartialWithdrawal]
```
These require a limit field.

dapplion replied:
For the `pending_balance_deposits`, a theoretical max would be the entire ETH supply as 1 ETH deposits, so ~120M ~ 2**27. How many more multiples of safety are needed?

For `pending_partial_withdrawals`, those will be bounded by economical fees on the EL layer. If that bound matches the inclusion churn on CL they should not backup infinitely. Unsure what number would make sense here, other than an upper bound like for `pending_balance_deposits`

rolfyone replied:
I'm missing why we need to start storing `pending_partial_withdrawals` now, this seems a step backwards from capella...

mkalinin replied:
Messages initiating partial withdrawals are coming from the EL without any hard limit on a number of such messages and the queue allows to have the limitation

potuz commented on specs/_features/maxeb_increase/capella.py:
```diff
@@ -1787,16 +1887,15 @@ def apply_deposit(state: BeaconState,
         if bls.Verify(pubkey, signing_root, signature):
-            state.validators.append(get_validator_from_deposit(pubkey, withdrawal_credentials, amount))
-            state.balances.append(amount)
+            state.validators.append(get_validator_from_deposit(pubkey, withdrawal_credentials))
```
What is this diff based on?

potuz replied:
You need to set index in this branch

dapplion replied:
It's based on the bundling of all specs of capella and previous at some commit ~1 month old. To generate the source file run `make pyspec` on the repo's root https://github.com/ethereum/consensus-specs/blob/e3a939e439d6c05356c9c29c5cd347384180bc01/Makefile#L104

potuz commented on specs/_features/maxeb_increase/capella.py:
```diff
@@ -1681,6 +1734,47 @@ def for_ops(operations: Sequence[Any], fn: Callable[[BeaconState, Any], None]) -
+    is_execution_address = validator.withdrawal_credentials[:1] == ETH1_ADDRESS_WITHDRAWAL_PREFIX
+    is_correct_source_address = validator.withdrawal_credentials[12:] == execution_layer_withdraw_request.source_address
+    if not (is_execution_address and is_correct_source_address):
+        return
```
I think it's better to assert here and declare the block as invalid.

mkalinin replied:
It should be a noop rather than assert to skip invalid withdrawal/exist request as EL can't know whether a request is invalid and will keep including it in the next block.

potuz replied:
I'd had assumed that we would have the CL not packing these rather than keeping these objects indefinitely in the mempool with the EL trying to pack them every block

mkalinin replied:
The entire solution involves a queue in the EL's world state elements of which are dequeued every built block, block invalidation instead of a no op would lead a system into an indefinite loop.

potuz reviewed (commented):
Do you want to include the `BeaconBlockBody(Container)` changes here as well?

potuz reviewed (commented):
depending on the MAX_EFFECTIVE_BALANCE constnant chosen, we need to change also the implementation of `compute_proposer_index` to use something more granular than a random byte that would not give enough granularity.

fradamt commented on specs/_features/maxeb_increase/capella.py:
```diff
@@ -1134,6 +1161,25 @@ def decrease_balance(state: BeaconState, index: ValidatorIndex, delta: Gwei) ->
+    else: # Exit doesn't fit in the current earliest epoch.
+        balance_to_process = exit_balance - state.exit_balance_to_consume
+        additional_epochs, remainder = divmod(balance_to_process, per_epoch_churn)
+        state.earliest_exit_epoch += additional_epochs
```
```suggestion
        state.earliest_exit_epoch += additional_epochs + 1
```
For example, if state.exit_balance_to_consume = exit_balance (we just fill the current epoch and there's no remaining balance to process) we still want to increase state.earliest_exit_epoch by 1

mkalinin replied:
Good catch!

But handling this case is a bit more complicated:
if state.exit_balance_to_consume == exit_balance then we want to return state.earliest_exit_epoch and only after that increase it by 1.

Making a respective change to the `if` condition above: replace `<` with `<=` should do the trick.

fradamt commented on specs/_features/maxeb_increase/capella.py:
```diff
@@ -1316,7 +1359,7 @@ def get_unslashed_attesting_indices(state: BeaconState,
     for a in attestations:
         output = output.union(get_attesting_indices(state, a.data, a.aggregation_bits))
-    return set(filter(lambda index: not state.validators[index].slashed, output))
+    return set(filter(lambda index: not is_slashed_attester(state.validators[index]), output))
```
I think we should have `is_slashed_validator(validator): return is_slashed_attester(validator) or is_slashed_proposer(validator)`, and use that here. Don't see why attestations from someone slashed as a proposer should count here

fradamt commented on specs/_features/maxeb_increase/capella.py:
```diff
@@ -1402,7 +1445,7 @@ def get_eligible_validator_indices(state: BeaconState) -> Sequence[ValidatorInde
     return [
         ValidatorIndex(index) for index, v in enumerate(state.validators)
-        if is_active_validator(v, previous_epoch) or (v.slashed and previous_epoch + 1 < v.withdrawable_epoch)
+        if is_active_validator(v, previous_epoch) or (is_slashed_attester(v) and previous_epoch + 1 < v.withdrawable_epoch)
```
Same here, would replace with this:

```suggestion
        if is_active_validator(v, previous_epoch) or (is_slashed_validator(v) and previous_epoch + 1 < v.withdrawable_epoch)
```

fradamt commented on specs/_features/maxeb_increase/capella.py:
```diff
@@ -1316,7 +1359,7 @@ def get_unslashed_attesting_indices(state: BeaconState,
     for a in attestations:
         output = output.union(get_attesting_indices(state, a.data, a.aggregation_bits))
-    return set(filter(lambda index: not state.validators[index].slashed, output))
+    return set(filter(lambda index: not is_slashed_attester(state.validators[index]), output))
```
I think we should have `is_slashed_validator(validator): return is_slashed_attester(validator) or is_slashed_proposer(validator)`, and use that here. Don't see why attestations from someone slashed as a proposer should count here

```suggestion
    return set(filter(lambda index: not is_slashed_validator(state.validators[index]), output))
```

mkalinin replied:
The intention of decoupling proposer and attester slashing states was to prevent proposer from suffering from attestation and inactivity leak penalties as well as the correlation penalty. So a proposer pays a fixed penalty for equivocation.

Both, `get_unslashed_attesting_indices` and `get_eligible_validator_indices` are used in computing those penalties hence the change in these functions. A way around that is to add `is_slashed_proposer` into specific places where penalties are calculated.

If we’re about to remove initial slashing penalty in general, maybe we can revert the decoupling slashings change to avoid unnecessary complexity.

fradamt replied:
I see, didn’t realize they were used to calculate penalties as well. I think adding it into specific places would be preferable

fradamt commented on specs/_features/maxeb_increase/capella.py:
```diff
@@ -2864,6 +2968,14 @@ def has_flag(flags: ParticipationFlags, flag_index: int) -> bool:
+def is_slashed_attester(validator: Validator) -> bool:
+    return has_flag(ParticipationFlags(validator.slashed), SLASHED_ATTESTER_FLAG_INDEX)
+
+
```
```suggestion
def is_slashed_validator(validator: Validator) -> bool:
    return is_slashed_attester(validator) or is_slashed_proposer(validator)
```

mkalinin commented on specs/_features/maxeb_increase/capella.py:
```diff
@@ -1134,6 +1161,25 @@ def decrease_balance(state: BeaconState, index: ValidatorIndex, delta: Gwei) ->
+        state.exit_balance_to_consume = per_epoch_churn
+
+    # Exit fits in the current earliest epoch.
+    if exit_balance < state.exit_balance_to_consume:
```
```suggestion
    if exit_balance <= state.exit_balance_to_consume:
```

See [the comment](https://github.com/michaelneuder/consensus-specs/pull/3/files#r1424874764) below

potuz reviewed (commented):
I think in `compute_proposer_index` it may be necessary to replace     `MAX_RANDOM_BYTE = 2**8 - 1` by a larger integer type to get more granularity.

mkalinin committed:
Apply suggestions from code review

Co-authored-by: fradamt <104826920+fradamt@users.noreply.github.com>

mkalinin commented on specs/_features/maxeb_increase/capella.py:
```diff
@@ -1681,6 +1734,47 @@ def for_ops(operations: Sequence[Any], fn: Callable[[BeaconState, Any], None]) -
+
+    # Same conditions as in EIP7002 https://github.com/ethereum/consensus-specs/pull/3349/files#diff-7a6e2ba480d22d8bd035bd88ca91358456caf9d7c2d48a74e1e900fe63d5c4f8R223
+    # Verify withdrawal credentials
+    is_execution_address = validator.withdrawal_credentials[:1] == ETH1_ADDRESS_WITHDRAWAL_PREFIX
```
```suggestion
    is_execution_address = validator.withdrawal_credentials[:1] == COMPOUNDING_WITHDRAWAL_PREFIX
```

mkalinin commented on specs/_features/maxeb_increase/capella.py:
```diff
@@ -1144,11 +1190,7 @@ def initiate_validator_exit(state: BeaconState, index: ValidatorIndex) -> None:
-    exit_queue_churn = len([v for v in state.validators if v.exit_epoch == exit_queue_epoch])
-    if exit_queue_churn >= get_validator_churn_limit(state):
-        exit_queue_epoch += Epoch(1)
+    exit_queue_epoch = compute_exit_epoch_and_update_churn(state, state.balances[index])
```
```suggestion
    exit_queue_epoch = compute_exit_epoch_and_update_churn(state, validator.effective_balance)
```

Pointed by @fradamt, `balances[index]` from the `effective_balance`, exit churn should rather be based on the latter.

fradamt committed:
consolidation by moving balance upon exit

mkalinin committed:
Revert "Constant penalty for proposer equivocations"

michaelneuder committed:
Merge pull request #15 from michaelneuder/revert-7-proposer-equivocation-penalty

Revert "Constant penalty for proposer equivocations"

fradamt committed:
separate consolidation churn from exit churn

mkalinin committed:
Make initial slashing penalty negligible

fradamt committed:
Set consolidation churn to leftover from activation and exit churn

mkalinin committed:
Update specs/_features/maxeb_increase/capella.py

Co-authored-by: fradamt <104826920+fradamt@users.noreply.github.com>

mkalinin committed:
Update specs/_features/maxeb_increase/capella.py

Co-authored-by: fradamt <104826920+fradamt@users.noreply.github.com>

mkalinin committed:
Set MAX_CONSOLIDATIONS to 1

mkalinin committed:
Polish process_pending_consolidations

mkalinin committed:
Merge pull request #14 from fradamt/consolidate-through-exit

consolidation by moving balance upon exit

fradamt committed:
fix issue in apply_deposit

`index` was only set in the case of a top-up and not of a deposit

mkalinin committed:
Merge branch 'maxeb-pyspec-min-viable' into mkalinin-patch-2

fradamt committed:
small fixes

off by one index and need keyword arguments

mkalinin committed:
Merge pull request #16 from michaelneuder/mkalinin-patch-2

Make initial slashing penalty negligible

fradamt commented on specs/_features/maxeb_increase/capella.py:
```diff
@@ -1840,6 +2014,40 @@ def process_voluntary_exit(state: BeaconState, signed_voluntary_exit: SignedVolu
+    target_validator = state.validators[consolidation.target_index]
+    source_validator = state.validators[consolidation.source_index]
+    # Verify the source and the target are active
+    assert is_active_validator(source_validator)
+    assert is_active_validator(target_validator)
```
```suggestion
    assert is_active_validator(source_validator, get_current_epoch(state))
    assert is_active_validator(target_validator, get_current_epoch(state))
```

fradamt commented on specs/_features/maxeb_increase/capella.py:
```diff
@@ -1575,18 +1663,63 @@ def process_eth1_data_reset(state: BeaconState) -> None:
+    state.pending_balance_deposits = state.pending_balance_deposits[next_pending_deposit_index:]
+
+
+def get_active_balance(state: BeaconState, validator: Validator) -> Gwei:
+    active_balance_ceil = MIN_ACTIVATION_BALANCE if has_eth1_withdrawal_credential(validator) else MAX_EFFECTIVE_BALANCE
+    return min(state.balances[validator.index], active_balance_ceil)
```
```suggestion
def get_active_balance(state: BeaconState, validator_index: ValidatorIndex) -> Gwei:
    active_balance_ceil = MIN_ACTIVATION_BALANCE if has_eth1_withdrawal_credential(state.validators[validator_index]) else MAX_EFFECTIVE_BALANCE
    return min(state.balances[validator_index], active_balance_ceil)
```

fradamt commented on specs/_features/maxeb_increase/capella.py:
```diff
@@ -1575,18 +1663,63 @@ def process_eth1_data_reset(state: BeaconState) -> None:
+    source_validator = state.validators[pending_consolidation.source_index]
+    target_validator = state.validators[pending_consolidation.target_index]
+    # Move active balance to target. Excess balance will be withdrawn.
+    active_balance = get_active_balance(state, source_validator)
```
```suggestion
    active_balance = get_active_balance(state, source_index)
```

fradamt commented on specs/_features/maxeb_increase/capella.py:
```diff
@@ -1575,18 +1663,63 @@ def process_eth1_data_reset(state: BeaconState) -> None:
+    return min(state.balances[validator.index], active_balance_ceil)
+
+def apply_pending_consolidation(state: BeaconState, pending_consolidation: PendingConsolidation) -> None:
+    source_validator = state.validators[pending_consolidation.source_index]
+    target_validator = state.validators[pending_consolidation.target_index]
+    # Move active balance to target. Excess balance will be withdrawn.
+    active_balance = get_active_balance(state, source_validator)
+    state.balances[source_validator.index] -= active_balance
+    state.balances[target_validator.index] += active_balance
```
```suggestion
    # Move active balance to target. Excess balance will be withdrawn.
    active_balance = get_active_balance(state, pending_consolidation.source_index)
    state.balances[pending_consolidation.source_index] -= active_balance
    state.balances[pending_consolidation.target_index] += active_balance
```

fradamt commented on specs/_features/maxeb_increase/capella.py:
```diff
@@ -1840,6 +2014,40 @@ def process_voluntary_exit(state: BeaconState, signed_voluntary_exit: SignedVolu
+    assert bls.FastAggregateVerify(pubkeys, signing_root, signed_consolidation.signature)
+
+    # Initiate source validator exit and append pending consolidation
+    active_balance = get_active_balance(state, source_validator)
```
```suggestion
    active_balance = get_active_balance(state, consolidation.source_index)
```

fradamt commented on specs/_features/maxeb_increase/capella.py:
```diff
@@ -1840,6 +2014,40 @@ def process_voluntary_exit(state: BeaconState, signed_voluntary_exit: SignedVolu
+    active_balance = get_active_balance(state, source_validator)
+    source_validator.exit_epoch = compute_consolidation_epoch_and_update_churn(state, active_balance)
+    source_validator.withdrawable_epoch = Epoch(source_validator.exit_epoch + config.MIN_VALIDATOR_WITHDRAWABILITY_DELAY)
+    state.pending_consolidations.append(PendingConsolidation(source_index = source_validator.index,
+                                                             target_index = target_validator.index))
```
```suggestion
    state.pending_consolidations.append(PendingConsolidation(source_index = consolidation.source_index,
                                                             target_index = consolidation.target_index))
```

mkalinin committed:
Apply suggestions from code review

Co-authored-by: fradamt <104826920+fradamt@users.noreply.github.com>

mkalinin committed:
Update specs/_features/maxeb_increase/capella.py

dapplion commented on specs/_features/maxeb_increase/capella.py:
```diff
@@ -2334,9 +2540,14 @@ def get_slot_signature(state: BeaconState, slot: Slot, privkey: int) -> BLSSigna
+    min_balance_increments = validator.effective_balance // MIN_ACTIVATION_BALANCE
+    committee_balance_increments = get_total_balance(state, set(committee)) // MIN_ACTIVATION_BALANCE
+    denominator = committee_balance_increments ** min_balance_increments
+    numerator = denominator - (committee_balance_increments -  TARGET_AGGREGATORS_PER_COMMITTEE) ** min_balance_increments
```
This intermediary product can overflow u64: a committee with 1M active indexes is 488 in size. If all members are consolidated committee_balance_increments = 31232, and min_balance_increments max value is 64. So the intermediate exponentiation is (488 * 64) ** 64 = 4e+287. Even if a single committee member is consolidated, (488 * 1) ** 64 = 1e172

EDIT: 2nd issue: If `validator. effective_balance < 32` this code attempt to divide by zero.

dapplion replied:
I _think_ (have not double checked) that this code achieves the same probabilities, without dividing by zero or having large intermediary values

```python
factor = total_committee_balance // validator_balance
modulo = max(1, factor // TARGET_AGGREGATORS_PER_COMMITTEE)
```

It's conceptually equivalent to the code before where:
```python
validator_indexes = 1
total_committee_indexes = len(get_beacon_committee(state, slot, index))
factor = total_committee_indexes / validator_indexes
modulo = max(1, factor // TARGET_AGGREGATORS_PER_COMMITTEE)
```

fradamt replied:
I had been thinking to simplify it to this, just explicitly looping through each "virtual validator" (corresponding to 32 eth) and checking if it is an aggregator roughly the same way we do today:

```python
def is_aggregator(state: BeaconState, slot: Slot, index: CommitteeIndex, validator_index: ValidatorIndex, slot_signature: BLSSignature) -> bool:
    validator = state.validators[validator_index]
    committee = get_beacon_committee(state, slot, index)
    virtual_validators = max(1, state.balances[validator] // MIN_ACTIVATION_BALANCE)
    modulo = max(1, len(committee) // TARGET_AGGREGATORS_PER_COMMITTEE)
return any(is_aggregator_virtual_validator(slot_signature, i, modulo) for i in range(virtual_validators))

def is_aggregator_virtual_validator(slot_signature: BLSSignature, virtual_validator_index: ValidatorIndex, modulo: uint64) -> bool:
    vrf_bytes = hash(uint_to_bytes(bytes_to_uint64(slot_signature) + virtual_validator_index))
    return bytes_to_uint64(vrf_bytes[0:8]) % modulo == 0
```

The difference between this approach (which is equivalent to the weird one that's currently in this spec) and what you're proposing is that your approach still targets 16 validators as aggregators, whereas this approach targets 16 "virtual validators" as aggregators, which on average would correspond to less than 16 aggregators if there is consolidation, because multiple virtual validators might map to the same validator. Originally I had chosen this approach with virtual validators because it seems to map exactly to what we have today, but I don't really see a reason why your approach wouldn't be fine as well. At the end of the day what's important is that:
- there's not too many aggregators
- there's high probability of at least an honest aggregator

Will give it more thought, but I think I would go with your approach :)

avsetsin commented on specs/_features/maxeb_increase/capella.py:
```diff
@@ -1840,6 +2012,40 @@ def process_voluntary_exit(state: BeaconState, signed_voluntary_exit: SignedVolu
 
+def process_consolidation(state: BeaconState, signed_consolidation: SignedConsolidation) -> None:
+    assert(len(state.pending_consolidations) < PENDING_CONSOLIDATIONS_LIMIT)
+    consolidation = signed_consolidation.message
+    target_validator = state.validators[consolidation.target_index]
+    source_validator = state.validators[consolidation.source_index]
```
I guess it's worth making sure that these are two different validators
```suggestion
    consolidation = signed_consolidation.message
    assert(consolidation.target_index != consolidation.source_index)
    target_validator = state.validators[consolidation.target_index]
    source_validator = state.validators[consolidation.source_index]
```

dapplion replied:
Checked in the canonical PR https://github.com/ethereum/consensus-specs/pull/3618

avsetsin commented on specs/_features/maxeb_increase/capella.py:
```diff
@@ -3540,24 +3751,41 @@ def has_eth1_withdrawal_credential(validator: Validator) -> bool:
     """
     Check if ``validator`` is partially withdrawable.
     """
-    has_max_effective_balance = validator.effective_balance == MAX_EFFECTIVE_BALANCE
```
It looks like the check `has_max_effective_balance` was removed by mistake.

There may be a situation where a validator has lost 0.25 eth below `MAX_EFFECTIVE_BALANCE` resulting in its effective balance decreasing by 1 eth. The validator could then bring the balance up to `MAX_EFFECTIVE_BALANCE` or higher, but less than `MAX_EFFECTIVE_BALANCE + 0.25 eth` (the value needed to return the effective balance to `MAX_EFFECTIVE_BALANCE`). 

The current code allows partial withdrawals of the validator balance when `effective balance < MAX_EFFECTIVE_BALANCE`, because of which it is very likely that the validator balance will not reach the necessary `MAX_EFFECTIVE_BALANCE + 0.25 eth` to return its effective balance to equal `MAX_EFFECTIVE_BALANCE`.

The same is true for 0x01 type and `MIN_ACTIVATION_BALANCE`

```python
def is_partially_withdrawable_validator(validator: Validator, balance: Gwei) -> bool:
    """
    Check if ``validator`` is partially withdrawable.
    """
    if has_eth1_withdrawal_credential(validator) and validator.effective_balance == MIN_ACTIVATION_BALANCE:
        return get_validator_excess_balance(validator, balance) > 0
    if has_compounding_withdrawal_credential(validator) and validator.effective_balance == MAX_EFFECTIVE_BALANCE:
        return get_validator_excess_balance(validator, balance) > 0
    return False
```

dapplion replied:
@avsetsin Please can you port your review comments to https://github.com/ethereum/consensus-specs/pull/3618 ?

dapplion commented:
Closing for
- https://github.com/ethereum/consensus-specs/pull/3618
